<html>
<head>
<title>Errors and exception handling in Python</title>
<link href="../Styles/Style.css" type="text/css" rel="stylesheet"/>
</head>
<body>
<h1 class="translated">Python中的错误和异常处理</h1>
<blockquote>原文：<a href="https://levelup.gitconnected.com/errors-and-exception-handling-in-python-e3560a879119?source=collection_archive---------13-----------------------#2020-10-15">https://levelup.gitconnected.com/errors-and-exception-handling-in-python-e3560a879119?source=collection_archive---------13-----------------------#2020-10-15</a></blockquote><div><div class="fc ie if ig ih ii"/><div class="ij ik il im in"><div class=""/><div class=""><h2 id="3cb1" class="pw-subtitle-paragraph jn ip iq bd b jo jp jq jr js jt ju jv jw jx jy jz ka kb kc kd ke dk translated">学习Python中错误和异常处理的指南</h2></div><figure class="kg kh ki kj gt kk gh gi paragraph-image"><div role="button" tabindex="0" class="kl km di kn bf ko"><div class="gh gi kf"><img src="../Images/bdbbcbb690156721f8d0388cd88f477e.png" data-original-src="https://miro.medium.com/v2/resize:fit:1400/0*AoyCENq56pED3qvq"/></div></div><figcaption class="kr ks gj gh gi kt ku bd b be z dk translated">穆斯塔法·梅拉吉在<a class="ae kv" href="https://unsplash.com?utm_source=medium&amp;utm_medium=referral" rel="noopener ugc nofollow" target="_blank"> Unsplash </a>上的照片</figcaption></figure><p id="66b9" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">编写代码时，错误是必然会发生的，但是可以进行错误处理来为这些错误做准备。通常，当出现错误时，脚本会停止工作。但是错误处理可以用来让脚本继续运行，即使有错误。</p><p id="c217" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">为此，我们需要了解3个关键词。它们如下:</p><p id="17b8" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated"><strong class="ky ir"> 1)试试</strong></p><p id="f49d" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">这是要运行的代码块，可能会导致错误</p><p id="09dd" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated"><strong class="ky ir"> 2)除:</strong></p><p id="6e65" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">如果try块中出现错误，将执行这段代码。</p><p id="e8c6" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated"><strong class="ky ir"> 3)最后:</strong></p><p id="3d83" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">不管有没有错误，这段代码都将被执行。</p><p id="bff9" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">所以让我们假设一个函数“求和”,它将两个数相加。该功能运行正常。</p><pre class="kg kh ki kj gt ls lt lu lv aw lw bi"><span id="d348" class="lx ly iq lt b gy lz ma l mb mc"><strong class="lt ir">&gt;&gt;&gt; def</strong> summation(num1,num2):<br/>        print(num1<strong class="lt ir">+</strong>num2)</span><span id="031e" class="lx ly iq lt b gy md ma l mb mc">&gt;&gt;&gt; summation(2,3)<br/>5</span></pre><p id="c2c4" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">现在让我们从用户那里获取一个输入，然后运行这个函数。</p><pre class="kg kh ki kj gt ls lt lu lv aw lw bi"><span id="b953" class="lx ly iq lt b gy lz ma l mb mc">&gt;&gt;&gt; num1 <strong class="lt ir">=</strong> 2<br/>&gt;&gt;&gt; num2 <strong class="lt ir">=</strong> input("Enter number: ")<br/>Enter number: 3</span><span id="f827" class="lx ly iq lt b gy md ma l mb mc">&gt;&gt;&gt; summation(num1,num2)</span><span id="cc91" class="lx ly iq lt b gy md ma l mb mc">&gt;&gt;&gt; print("This line will not be printed because of the error")<br/><strong class="lt ir">---------------------------------------------------------------------------</strong><br/><strong class="lt ir">TypeError</strong>                                 Traceback (most recent call last)<br/><strong class="lt ir">&lt;ipython-input-6-2cc0289b921e&gt;</strong> in &lt;module&gt;<br/><strong class="lt ir">----&gt; 1 </strong>summation<strong class="lt ir">(</strong>num1<strong class="lt ir">,</strong>num2<strong class="lt ir">)</strong><br/>      2 print<strong class="lt ir">("This line will not be printed because of the error")</strong><br/><br/><strong class="lt ir">&lt;ipython-input-1-970d26ae8592&gt;</strong> in summation<strong class="lt ir">(num1, num2)</strong><br/>      1 <strong class="lt ir">def</strong> summation<strong class="lt ir">(</strong>num1<strong class="lt ir">,</strong>num2<strong class="lt ir">):</strong><br/><strong class="lt ir">----&gt; 2     </strong>print<strong class="lt ir">(</strong>num1<strong class="lt ir">+</strong>num2<strong class="lt ir">)</strong><br/><br/><strong class="lt ir">TypeError</strong>: unsupported operand type(s) for +: 'int' and 'str'</span></pre><p id="e7bd" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">所以我们得到一个错误“TypeError ”,因为我们试图添加一个数字和一个字符串。请注意，错误发生后，后面的代码行将不会被执行。所以我们使用上面提到的关键字来确保即使发生错误，脚本也能运行。</p><pre class="kg kh ki kj gt ls lt lu lv aw lw bi"><span id="b84f" class="lx ly iq lt b gy lz ma l mb mc"><strong class="lt ir">&gt;&gt;&gt; try</strong>:<br/>        summed <strong class="lt ir">=</strong> 2 <strong class="lt ir">+</strong> '3'<br/>    <strong class="lt ir">except</strong>:<br/>        print("Summation is not of the same type")</span><span id="3943" class="lx ly iq lt b gy md ma l mb mc">Summation is not of the same type</span></pre><p id="0f3b" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">所以在这里我们可以看到，当try块中出现错误时，它会转到except并打印语句。现在让我们添加“else”块，以防没有遇到错误。</p><pre class="kg kh ki kj gt ls lt lu lv aw lw bi"><span id="0ad9" class="lx ly iq lt b gy lz ma l mb mc"><strong class="lt ir">&gt;&gt;&gt; try</strong>:<br/>        summed <strong class="lt ir">=</strong> 2 <strong class="lt ir">+</strong> 3<br/>    <strong class="lt ir">except</strong>:<br/>        print("Summation is not of the same type")<br/>    <strong class="lt ir">else</strong>:<br/>        print("There was no error and result is: ",summed)</span><span id="009e" class="lx ly iq lt b gy md ma l mb mc">There was no error and result is:  5</span></pre><p id="ca65" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">现在让我们用另一个例子来理解这一点，在这个例子中，我们也提到了except块的错误。如果你没有在except块中提到错误类型，它将为所有的异常假定错误类型。</p><pre class="kg kh ki kj gt ls lt lu lv aw lw bi"><span id="6b03" class="lx ly iq lt b gy lz ma l mb mc"><strong class="lt ir">&gt;&gt;&gt; try</strong>:<br/>        f <strong class="lt ir">=</strong> open('test','w')<br/>        f.write("This is a test file")<br/>    <strong class="lt ir">except</strong> TypeError:<br/>        print("There is a type error")<br/>    <strong class="lt ir">except</strong> OSError:<br/>        print("There is an OS error")<br/>    <strong class="lt ir">finally</strong>:<br/>        print("This will print even if no error")</span><span id="29ed" class="lx ly iq lt b gy md ma l mb mc">This will print even if no error</span></pre><p id="d70a" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">现在创建一个错误来查看except块是否与finally块一起工作</p><pre class="kg kh ki kj gt ls lt lu lv aw lw bi"><span id="b891" class="lx ly iq lt b gy lz ma l mb mc"><strong class="lt ir">&gt;&gt;&gt; try</strong>:<br/>        f <strong class="lt ir">=</strong> open('test','r')<br/>        f.write("This is a test file")<br/>    <strong class="lt ir">except</strong> TypeError:<br/>        print("There is a type error")<br/>    <strong class="lt ir">except</strong> OSError:<br/>        print("There is an OS error")<br/>    <strong class="lt ir">finally</strong>:<br/>        print("This will print even if no error")</span><span id="b0c5" class="lx ly iq lt b gy md ma l mb mc">There is an OS error<br/>This will print even if no error</span></pre><p id="d283" class="pw-post-body-paragraph kw kx iq ky b kz la jr lb lc ld ju le lf lg lh li lj lk ll lm ln lo lp lq lr ij bi translated">参考笔记本<a class="ae kv" href="https://github.com/jayashree8/Python_guide/blob/master/Python%20errors%20and%20exception%20handling/Errors%20and%20exception%20handling%20in%C2%A0Python.ipynb" rel="noopener ugc nofollow" target="_blank">此处</a>。</p><h2 id="72c7" class="lx ly iq bd me mf mg dn mh mi mj dp mk lf ml mm mn lj mo mp mq ln mr ms mt mu bi translated">学习Python可以参考的入门书籍:</h2><div class="mv mw gp gr mx my"><a href="https://amzn.to/3yDY4To" rel="noopener  ugc nofollow" target="_blank"><div class="mz ab fo"><div class="na ab nb cl cj nc"><h2 class="bd ir gy z fp nd fr fs ne fu fw ip bi translated">Python速成班，第二版:基于项目的编程入门实践</h2><div class="nf l"><h3 class="bd b gy z fp nd fr fs ne fu fw dk translated">世界上最畅销的Python书籍的第二版。一个快速的，没有废话的Python编程指南…</h3></div></div><div class="ng l"><div class="nh l ni nj nk ng nl kp my"/></div></div></a></div><div class="mv mw gp gr mx my"><a href="https://amzn.to/3vtvQZv" rel="noopener  ugc nofollow" target="_blank"><div class="mz ab fo"><div class="na ab nb cl cj nc"><h2 class="bd ir gy z fp nd fr fs ne fu fw ip bi translated">艰难地学习Python:一个非常简单的介绍可怕的美丽世界…</h2><div class="nf l"><h3 class="bd b gy z fp nd fr fs ne fu fw dk translated">你会学习Python！Zed Shaw完善了世界上最好的学习Python的系统。遵循它，你会…</h3></div></div><div class="ng l"><div class="nm l ni nj nk ng nl kp my"/></div></div></a></div><div class="mv mw gp gr mx my"><a href="https://amzn.to/3urluYI" rel="noopener  ugc nofollow" target="_blank"><div class="mz ab fo"><div class="na ab nb cl cj nc"><h2 class="bd ir gy z fp nd fr fs ne fu fw ip bi translated">思考Python，2e:如何像计算机科学家一样思考</h2><div class="nf l"><h3 class="bd b gy z fp nd fr fs ne fu fw dk translated">思考Python，2e:如何像计算机科学家一样思考</h3></div></div><div class="ng l"><div class="nn l ni nj nk ng nl kp my"/></div></div></a></div><h2 id="c31d" class="lx ly iq bd me mf mg dn mh mi mj dp mk lf ml mm mn lj mo mp mq ln mr ms mt mu bi translated">学习Python可以参考的高级书籍:</h2><div class="mv mw gp gr mx my"><a href="https://amzn.to/3fMzVBn" rel="noopener  ugc nofollow" target="_blank"><div class="mz ab fo"><div class="na ab nb cl cj nc"><h2 class="bd ir gy z fp nd fr fs ne fu fw ip bi translated">编程Python:强大的面向对象编程</h2><div class="nf l"><h3 class="bd b gy z fp nd fr fs ne fu fw dk translated">如果你已经掌握了Python的基础，你就可以开始使用它来完成真正的工作了。编程Python将…</h3></div></div><div class="ng l"><div class="no l ni nj nk ng nl kp my"/></div></div></a></div><div class="mv mw gp gr mx my"><a href="https://amzn.to/34oFFMl" rel="noopener  ugc nofollow" target="_blank"><div class="mz ab fo"><div class="na ab nb cl cj nc"><h2 class="bd ir gy z fp nd fr fs ne fu fw ip bi translated">高级Python编程:使用以下工具构建高性能、并发和多线程应用</h2><div class="nf l"><h3 class="bd b gy z fp nd fr fs ne fu fw dk translated">关键特性使用Dask和PySpark Master技能在集群上设置和运行分布式算法，以准确地…</h3></div></div><div class="ng l"><div class="np l ni nj nk ng nl kp my"/></div></div></a></div><blockquote class="nq nr ns"><p id="f13f" class="kw kx nt ky b kz la jr lb lc ld ju le nu lg lh li nv lk ll lm nw lo lp lq lr ij bi translated"><em class="iq">联系我:</em> <a class="ae kv" href="https://www.linkedin.com/in/jayashree-domala8/" rel="noopener ugc nofollow" target="_blank"> <em class="iq"> LinkedIn </em> </a></p><p id="834a" class="kw kx nt ky b kz la jr lb lc ld ju le nu lg lh li nv lk ll lm nw lo lp lq lr ij bi translated"><em class="iq">查看我的其他作品:</em> <a class="ae kv" href="https://github.com/jayashree8" rel="noopener ugc nofollow" target="_blank"> <em class="iq"> GitHub </em> </a></p></blockquote></div></div>    
</body>
</html>